go/vt/tabletserver/tx_engine.go

```diff
@@ -0,0 +1,252 @@
+		return
+	}
+
+	te.readPool.Open(&dbconfigs.App, &dbconfigs.Dba)
+	if err := te.prepareFromRedo(); err != nil {
+		// If this operation fails, we choose to raise an alert and
+		// continue anyway. Serving traffic is considered more important
+		// than blocking everything for the sake of a few transactions.
+		te.queryServiceStats.InternalErrors.Add("TwopcResurrection", 1)
+		log.Errorf("Could not prepare transactions: %v", err)
+	}
+	te.startWatchdog()
+	te.isOpen = true
+}
+
+// Close closes the TxEngine. If the immediate flag is on,
+// then all current transactions are immediately rolled back.
+// Otherwise, the function waits for all current transactions
+// to conclude.
+func (te *TxEngine) Close(immediate bool) {
+	if !te.isOpen {
+		return
+	}
+	if te.twopcEnabled {
+		te.stopWatchdog()
+		te.readPool.Close()
+	}
+	if immediate {
+		te.rollbackTransactions()
+	}
+	te.txPool.WaitForEmpty()
+	te.txPool.Close()
+	te.isOpen = false
+}
+
+// prepareFromRedo replays and prepares the transactions
+// from the redo log. It also loads previously failed transactions
+// into the reserved list.
+// TODO(sougou): Make this function set the lastId for tx pool to be
+// greater than all those used by dtids. This will prevent dtid
+// collisions.
+func (te *TxEngine) prepareFromRedo() error {
+	ctx := context.Background()
+	var allErr concurrency.AllErrorRecorder
+	conn, err := te.readPool.Get(ctx)
+	if err != nil {
+		return err
+	}
+	defer conn.Recycle()
+	prepared, failed, err := te.twoPC.ReadAllRedo(ctx, conn)
+	if err != nil {
+		return err
+	}
+
+outer:
+	for dtid, tx := range prepared {
+		conn, err := te.txPool.LocalBegin(ctx)
+		if err != nil {
+			allErr.RecordError(err)
+			continue
+		}
+		for _, stmt := range tx {
+			conn.RecordQuery(stmt)
+			_, err := conn.Exec(ctx, stmt, 1, false)
+			if err != nil {
+				allErr.RecordError(err)
+				te.txPool.LocalConclude(ctx, conn)
+				continue outer
+			}
+		}
+		// We should not use the external Prepare because
+		// we don't want to write again to the redo log.
+		err = te.preparedPool.Put(conn, dtid)
+		if err != nil {
+			allErr.RecordError(err)
+			continue
+		}
+	}
+	for _, dtid := range failed {
+		te.preparedPool.SetFailed(dtid)
+	}
+	return allErr.Error()
+}
+
+// rollbackTransactions rolls back all open transactions
+// including the prepared ones.
+// This is used for transitioning from a master to a non-master
+// serving type.
+func (te *TxEngine) rollbackTransactions() {
+	ctx := context.Background()
+	// The order of rollbacks is currently not material because
+	// we don't allow new statements or commits during
+	// this function. In case of any such change, this will
+	// have to be revisited.
+	te.txPool.RollbackNonBusy(ctx)
+	for _, c := range te.preparedPool.FetchAll() {
+		te.txPool.LocalConclude(ctx, c)
+	}
+}
+
+// startWatchdog starts the watchdog goroutine, which looks for abandoned
+// transactions and calls the notifier on them.
+func (te *TxEngine) startWatchdog() {
```

Comment:
That's what I started with. But it makes the shutdown sequence confusing, which should be watchdog->txPool->preparedPool->twoPC. Ideally, I think watchdog should be its own service under TxEngine. I kind of punted on this part.